### PR TITLE
Add is_public property to determine profile visibility. 404 on False.

### LINKED ISSUE #6042: make the accounts API 404 for non-developer users
..., except for admins and your own account
part of #5592

### LINKED ISSUE #6056: refine which users are "developers"; which users can see all user account profiles
part of #5592

Currently `UserProfile.is_developer` is defined as a user that has _any_ type of add-on associated with them, (owner, developer, viewer, 'listed' or not), that is not deleted, regardless of status.  Only users that are `is_developer` will have their profiles publicly viewable by all.

Users that will be able to see any user account profile (admin-ish) have `Users:Edit`.

Do either of these restrictions need to be changed?

### PR DESCRIPTION
**Pull request by eviljeff**
fixes #6042 and fixes #6056

### CODE REVIEW

File: docs/topics/api/accounts.rst [hunk 1]
```diff
@@ -12,8 +12,13 @@ Account
 .. _`account`:
 
 This endpoint returns information about a user's account, by the account id.
+Only :ref:`developer <developer_account>` accounts are publicly viewable - other user's accounts will return a 404 not found response code.
 Most of the information is optional and provided by the user so may be missing or inaccurate.
 
+.. _`developer_account`:
+A developer is defined as a user who is listed as a developer or owner of more approved add-ons.
```
Comment (EnTeQuAk) [inline docs/topics/api/accounts.rst]: "of one or more approved add-ons" sounds better?

File: src/olympia/accounts/views.py [hunk 2]
```diff
@@ -413,7 +413,13 @@ def get_object(self):
...
         self.kwargs[self.lookup_field] = identifier
         self.instance = super(AccountViewSet, self).get_object()
-        return self.instance
+        # action won't exist for other classes that are using this ViewSet.
+        action = getattr(self, 'action', None)
+        if (not action or self.self_view or self.admin_viewing or
+                self.instance.is_public):
+            return self.instance
+        else:
+            raise Http404
```
Comment (EnTeQuAk) [inline src/olympia/accounts/views.py]: I feel this is hard to read, I'd do something like

```python
can_view_instance = (
    not action or
    self.self_view or
    self.admin_viewing or
    self.instance.is_public)

if can_view_instance:
    return self.instance

raise Http404()

File: src/olympia/addons/models.py [hunk 3]
```diff
@@ -1420,6 +1421,10 @@ def watch_status(old_attr=None, new_attr=None, instance=None,
     latest_version = instance.find_latest_version(
         channel=amo.RELEASE_CHANNEL_LISTED)
 
+    # Update the author's account profile visibility
+    if new_status != old_status:
+        [author.update_is_public() for author in instance.authors.all()]
```
Comment (EnTeQuAk) [inline src/olympia/addons/models.py]: I wonder if this should happen in a celery task to not block for add-ons with many authors. Or just re-use `watch_addon_user` task?

File: src/olympia/migrations/965-add-user-is-public-and-drop-is-verified.sql [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+/* AUTHOR_ROLE_DEV = 4
+AUTHOR_ROLE_OWNER = 5
+STATUS_PUBLIC = 4 */
+
+UPDATE users, addons_users, addons SET users.`public`=true
+    WHERE users.`id`=addons_users.`user_id` and
+          addons_users.`role` in (4, 5) and
+          addons_users.`listed`=true and
+          addons_users.`addon_id`=addons.`id` and
+          addons.`status`=4;
```
Comment (EnTeQuAk) [inline src/olympia/migrations/965-add-user-is-public-and-drop-is-verified.sql]: `in` and `and` should be uppercase and please add whitespaces before and after `=`

File: src/olympia/migrations/965-add-user-is-public-and-drop-is-verified.sql [hunk 5]
```diff
@@ -0,0 +1,14 @@
+ALTER TABLE users
+    ADD COLUMN `public` boolean DEFAULT false,
```
Comment (EnTeQuAk) [inline src/olympia/migrations/965-add-user-is-public-and-drop-is-verified.sql]: `false` should be an uppercase keyword too

File: src/olympia/users/models.py [hunk 6]
```diff
@@ -142,7 +142,9 @@ class UserProfile(OnChangeMixin, ModelBase, AbstractBaseUser):
     failed_login_attempts = models.PositiveIntegerField(default=0,
                                                         editable=False)
 
-    is_verified = models.BooleanField(default=True)
+    # Is the profile page for this account publicly viewable?
+    # Note: this isn't applicable to legacy frontend pages.
```
Comment (EnTeQuAk) [inline src/olympia/users/models.py]: This needs a bit more explanation imho, why it's not or what this means for legacy frontend pages.

File: docs/topics/api/accounts.rst (hunk 1 above)
Comment (eviljeff) [inline docs/topics/api/accounts.rst]: 😒 I thought I wrote that already

File: src/olympia/addons/models.py (hunk 3 above)
Comment (eviljeff) [inline src/olympia/addons/models.py]: Not sure what you mean by reusing the `watch_addon_user` task - what task?

On a status update there is already a lot of things happening, including potentially files being moved and signed, so a couple of extra queries isn't going to be noticed.  Besides, there are vanishingly small numbers of add-ons with more than a handful (more than 1, really) of authors so we'd be handling an edge case that might not exist.

**Comment by eviljeff** [comment]
@EnTeQuAk https://github.com/mozilla/addons-server/pull/6219/commits/855b5bd2d2edc94556443ba0bc2396044b11f1ad

**Pull request merged**